js/uid2-sdk-3.0.0.ts

```diff
@@ -117,118 +113,123 @@ export class UID2 {
-                if (window.__uid2 && window.__uid2.getAdvertisingTokenAsync) {
-                    return window.__uid2.getAdvertisingTokenAsync();
-                } else {
-                    return Promise.reject(new Error("UID2 SDK not present"));
-                }
-            },
-        });
+    public callbacks: Uid2CallbackHandler[] = [];
+    constructor() {
+        this.runCallbacks(EventType.SdkLoaded, {});
+        this._tokenPromiseHandler = new UID2PromiseHandler();
+        this._tokenPromiseHandler.attachToSdk(this);
     }
 
-    public init(opts: UID2Options) {
+    public init(opts: Uid2Options) {
         this.initInternal(opts);
     }
-    public initInternal(opts: UID2Options | unknown) {
-        if (this._initCalled) {
+    public initInternal(opts: Uid2Options | unknown) {
+        if (this._initComplete) {
             throw new TypeError('Calling init() more than once is not allowed');
         }
-
         if (!isUID2OptionsOrThrow(opts)) throw new TypeError(`Options provided to UID2 init couldn't be validated.`);
         
         this._cookieManager = new UID2CookieManager({ ...opts, cookieName: UID2.COOKIE_NAME });
-        this._initCalled = true;
         this._opts = opts;
-
         this._apiClient = new Uid2ApiClient(this._opts.baseUrl ?? "https://prod.uidapi.com", 'uid2-sdk-' + UID2.VERSION);
 
-        const identity = this._opts.identity ? this._opts.identity : this.loadIdentity()
+        const identity = this._opts.identity ? this._opts.identity : this.loadIdentityFromCookie()
         this.applyIdentity(identity);
+        this._initComplete = true;
+        this.runCallbacks(EventType.InitCompleted, {});
     }
     public getAdvertisingToken() {
         return this._identity && !this.temporarilyUnavailable() ? this._identity.advertising_token : undefined;
     }
+    // If the SDK has been initialized, returns a resolved promise with the current token (or rejected if not available)
+    // Otherwise, returns a promise which will be resolved after init.
     public getAdvertisingTokenAsync() {
-        if (!this.initialised()) {
-            return new Promise((resolve, reject) => {
-                this._promises.push({ resolve: resolve, reject: reject });
-            });
-        } else if (this._identity) {
-            return this.temporarilyUnavailable()
-                ? Promise.reject(new Error('temporarily unavailable'))
-                : Promise.resolve(this._identity.advertising_token);
-        } else {
-            return Promise.reject(new Error('identity not available'));
-        }
+        const token = this.getAdvertisingToken();
+        return this._tokenPromiseHandler.createMaybeDeferredPromise(token ?? null);
     }
     public isLoginRequired() {
-        return this.initialised() ? !this._identity : undefined;
+        if (!this._initComplete) return undefined;
+        return !(this.isLoggedIn() || this._apiClient?.hasActiveRequests());
```

Comment:
Is `this._apiClient?.hasActiveRequests()` means there is a token generating or refreshing on going?